crates/dap/src/adapters.rs

```diff
@@ -244,49 +218,34 @@ pub trait DebugAdapter: 'static + Send + Sync {
         {
             log::info!("Using cached debug adapter binary {}", self.name());
 
-            return self.get_installed_binary(delegate, &config, None).await;
+            if let Some(binary) = self.get_installed_binary(delegate, &config).await.log_err() {
+                return Ok(binary);
+            }
+
+            log::info!(
+                "Cached binary {} is corrupt falling back to install",
+                self.name()
+            );
         }
 
         log::info!("Getting latest version of debug adapter {}", self.name());
         delegate.update_status(self.name(), DapStatus::CheckingForUpdate);
-        let version = self.fetch_latest_adapter_version(delegate).await.ok();
-
-        let mut binary = self.get_installed_binary(delegate, &config, None).await;
-
-        if let Some(version) = version {
-            if binary
-                .as_ref()
-                .is_ok_and(|binary| binary.version == version.tag_name)
-            {
-                delegate
-                    .updated_adapters()
-                    .lock_arc()
-                    .await
-                    .insert(self.name());
-
-                return binary;
-            }
-
+        if let Some(version) = self.fetch_latest_adapter_version(delegate).await.log_err() {
+            log::info!(
+                "Installiing latest version of debug adapter {}",
+                self.name()
+            );
             delegate.update_status(self.name(), DapStatus::Downloading);
             self.install_binary(version, delegate).await?;
 
-            binary = self.get_installed_binary(delegate, &config, None).await;
-        } else {
-            log::error!(
-                "Failed getting latest version of debug adapter {}",
-                self.name()
-            );
+            delegate
```

Comment:
I think we check if the binary is valid before adding it to updated_adapters the debug adapter

1. Fetching a version could fail while the binary is valid and this change doesn't cover that edge case
2. More importantly, we could be caching a binary that is invalid here which we should never do